source/garbageHandler.py

```diff
@@ -0,0 +1,71 @@
+# -*- coding: UTF-8 -*-
+# A part of NonVisual Desktop Access (NVDA)
+# Copyright (C) 2020 NV Access Limited
+# This file is covered by the GNU General Public License.
+# See the file COPYING for more details.
+
+
+import gc
+import threading
+from logHandler import log
+
+""" Watches Python's cyclic garbage collector and reports questionable collections. """
+
+
+class TrackedObject:
+	"""
+	An object that notifies garbageHandler when it is destructed,
+	so that NVDA can log when important unreachable objects are being deleted
+	by Python's cyclic garbage collector.
+	"""
+
+	def __del__(self):
+		notifyObjectDeletion(self)
+
+
+_collectionThreadID = 0
+_reportCountDuringCollection = 0
+
+
+def initialize():
+	""" Initializes NVDA's garbage handler. """
+	# Instruct Python to keep all unreachable objects for later inspection
+	# gc.set_debug(gc.DEBUG_SAVEALL)
+	# Register a callback with Python's garbage collector
+	# That will notify us of the start and end of each collection run.
+	gc.callbacks.append(_collectionCallback)
+
+
+def _collectionCallback(action, info):
+	global _collectionThreadID, _reportCountDuringCollection
+	if action == "start":
+		_collectionThreadID = threading.currentThread().ident
+		_reportCountDuringCollection = 0
+	elif action == "stop":
+		if _reportCountDuringCollection > 0:
+			log.error(f"Found at least {_reportCountDuringCollection} unreachable objects in run")

```

Comment:
It might be nice if this message said that it marked the end of this garbage collection run. Note that this error means that two errors will be logged if there are any object deletions.